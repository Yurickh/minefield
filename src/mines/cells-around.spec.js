import cellsAround from './cells-around'

function isSameCell({ row, col }) {
  return cell => row === cell.row && col === cell.col
}

function createSnapshotWithGrid({ numCols, numRows }) {
  return positions =>
    '\n' +
    Array(numRows)
      .fill()
      .map((_, row) =>
        Array(numCols)
          .fill()
          .map((_, col) =>
            positions.some(isSameCell({ row, col })) ? '#' : '.'
          )
          .join('')
      )
      .join('\n')
}

describe('cells-around', () => {
  let snapshot
  let getCellsAround

  beforeEach(() => {
    const grid = { numCols: 5, numRows: 5 }
    snapshot = createSnapshotWithGrid(grid)
    getCellsAround = cellsAround(grid)
  })

  it('is an array of cells', () => {
    expect(getCellsAround({ row: 3, col: 3 })).toEqual([
      { row: 2, col: 2 },
      { row: 2, col: 3 },
      { row: 2, col: 4 },
      { row: 3, col: 2 },
      { row: 3, col: 4 },
      { row: 4, col: 2 },
      { row: 4, col: 3 },
      { row: 4, col: 4 },
    ])
  })

  it('gets the cells around the center', () => {
    expect(snapshot(getCellsAround({ row: 2, col: 2 }))).toMatchSnapshot()
  })

  it('gets the cells around the corner', () => {
    expect(snapshot(getCellsAround({ row: 0, col: 0 }))).toMatchSnapshot()
    expect(snapshot(getCellsAround({ row: 4, col: 4 }))).toMatchSnapshot()
    expect(snapshot(getCellsAround({ row: 0, col: 4 }))).toMatchSnapshot()
    expect(snapshot(getCellsAround({ row: 4, col: 0 }))).toMatchSnapshot()
  })

  it('skips cells that overflow', () => {
    expect(getCellsAround({ row: 3, col: 3 })).toHaveLength(8)
    expect(getCellsAround({ row: 0, col: 0 })).toHaveLength(3)
  })
})
