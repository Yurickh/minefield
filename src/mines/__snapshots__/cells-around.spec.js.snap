// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cells-around gets the cells around the center 1`] = `
"
.....
.###.
.#.#.
.###.
....."
`;

exports[`cells-around gets the cells around the corner 1`] = `
"
.#...
##...
.....
.....
....."
`;

exports[`cells-around gets the cells around the corner 2`] = `
"
.....
.....
.....
...##
...#."
`;

exports[`cells-around gets the cells around the corner 3`] = `
"
...#.
...##
.....
.....
....."
`;

exports[`cells-around gets the cells around the corner 4`] = `
"
.....
.....
.....
##...
.#..."
`;
